src/Features/CSharp/Portable/CodeFixes/AddParameterXmlDocumentation/AddParameterXmlDocumentationCodeFixProvider.cs

```diff
@@ -0,0 +1,316 @@
+                {
+                    parameterIndices.Add(param, index);
+                    index++;
+                }
+                if (!parameterIndices.ContainsKey(parameterName))
+                {
+                    // something went very wrong
+                    Debug.Assert(false); // TODO message or remove assert
+                    adjacentNameAttribute = null;
+                    adjacentAttributeIsBefore = false;
+                    return;
+                }
+
+                var walker = new FindAdjacentParameterDocumentationWalker(parameterIndices[parameterName], parameterIndices, elementKind);
+                walker.VisitLeadingTrivia(tokenWithDocumentationComment);
+
+                if (walker._documentedParameterBefore != null)
+                {
+                    adjacentNameAttribute = walker._documentedParameterBefore;
+                    adjacentAttributeIsBefore = true;
+                }
+                else
+                {
+                    adjacentNameAttribute = walker._documentedParameterAfter;
+                    adjacentAttributeIsBefore = false;
+                }
+            }
+
+            public override void VisitXmlNameAttribute(XmlNameAttributeSyntax node)
+            {
+                if (node.GetElementKind() == _elementKind)
+                {
+                    int index;
+                    if (_parameterIndices.TryGetValue(node.Identifier.Identifier.ValueText, out index))
+                    {
+                        if (_documentedParameterBeforeIndex < index & index < _documentedParameterAfterIndex)
+                        {
+                            if (index < _parameterIndex)
+                            {
+                                _documentedParameterBefore = node;
+                                _documentedParameterBeforeIndex = index;
+                            }
+                            if (_parameterIndex < index)
+                            {
+                                _documentedParameterAfter = node;
+                                _documentedParameterAfterIndex = index;
+                            }
+                        }
+                    }
+                    else if (_documentedParameterAfter == null)
+                    {
+                        // in case all the param or typeparam tags that exist are not a parameter
+                        _documentedParameterAfter = node;
+                    }
+                }
+                base.VisitXmlNameAttribute(node);
+            }
+        }
+
+        private class AddParameterXmlDocumentationCodeAction : CodeAction
+        {
+            private string _title;
+            private Document _document;
+            private string _parameterName;
+            private XmlNameAttributeSyntax _adjacentAttribute;
+            private bool _adjacentAttributeIsBefore;
+            private bool _isTypeParameter;
```

Comment:
It doesn't looks like any of these fields are assigned outside of the constructor. Make them read only?